receiver/azureblobrehydrationreceiver/receiver.go

```diff
@@ -0,0 +1,417 @@
+					r.logger.Debug("Skipping Blob, non-matching blob path", zap.String("blob", blob.Name))
+				case err != nil:
+					r.logger.Error("Error processing blob path", zap.String("blob", blob.Name), zap.Error(err))
+				case checkpoint.ShouldParse(*blobTime, blob.Name):
+					// if the blob is not in the specified time range or not of the telemetry type supported by this receiver
+					// then skip consuming it.
+					if !r.isInTimeRange(*blobTime) || telemetryType != r.supportedTelemetry {
+						continue
+					}
+
+					// Process and consume the blob at the given path
+					if err := r.processBlob(blob); err != nil {
+						r.logger.Error("Error consuming blob", zap.String("blob", blob.Name), zap.Error(err))
+						continue
+					}
+
+					// Update and save the checkpoint with the most recently processed blob
+					checkpoint.UpdateCheckpoint(*blobTime, blob.Name)
+					if err := r.saveCheckpoint(r.ctx, checkpoint); err != nil {
+						r.logger.Error("Error while saving checkpoint", zap.Error(err))
+					}
+
+					// Delete blob if configured to do so
+					if r.cfg.DeleteOnRead {
+						if err := r.azureClient.DeleteBlob(r.ctx, r.cfg.Container, blob.Name); err != nil {
+							r.logger.Error("Error while attempting to delete blob", zap.String("blob", blob.Name), zap.Error(err))
+						}
+					}
+				}
+			}
+
+		}
+	}
+}
+
+// saveCheckpoint saves the checkpoint to keep place in rehydration effort
+func (r *rehydrationReceiver) saveCheckpoint(ctx context.Context, checkpoint *rehydrationCheckpoint) error {
+	data, err := json.Marshal(checkpoint)
+	if err != nil {
+		return fmt.Errorf("marshal checkpoint: %w", err)
+	}
+
+	return r.storageClient.Set(ctx, checkpointStorageKey, data)
+}
+
+// loadCheckpoint loads a checkpoint timestamp to be used to keep place in rehydration effort
+func (r *rehydrationReceiver) loadCheckpoint(ctx context.Context) *rehydrationCheckpoint {
+	checkpoint := newCheckpoint()
+
+	data, err := r.storageClient.Get(ctx, checkpointStorageKey)
+	if err != nil {
+		r.logger.Info("Unable to load checkpoint from storage client, continuing without a previous checkpoint", zap.Error(err))
+		return checkpoint
+	}
+
+	if data == nil {
+		return checkpoint
+	}
+
+	if err := json.Unmarshal(data, checkpoint); err != nil {
+		r.logger.Error("Error while decoding the stored checkpoint, continuing without a checkpoint", zap.Error(err))
+		return checkpoint
+	}
+
+	return checkpoint
+}
+
+// constants for blob path parts
+const (
+	year   = "year="
+	month  = "month="
+	day    = "day="
+	hour   = "hour="
+	minute = "minute="
+)
+
+// strings that indicate what type of telemetry is in a blob
+const (
+	metricBlobSignifier = "metrics_"
+	logsBlobSignifier   = "logs_"
+	tracesBlobSignifier = "traces_"
+)
+
+// parseBlobPath returns true if the blob is within the existing time range
+func parseBlobPath(prefix *string, blobName string) (blobTime *time.Time, telemetryType component.DataType, err error) {
```

Comment:
Just wondering if it's worth it to do all this state machine stuff over using regex and capture groups? Might be easier to understand.

Reply:
You are 100% I think I over engineered this problem. Let me look at using regexes.

Reply:
Ok so I started going the regex route and I found it was going to be almost exactly as complex code wise. The main thing I saw was parsing the timestamp. The regex gives me an array of capture groups that I have to iterate through and either construct a string like I am now  and pass it to `time.Parse` or do a bunch of string conversion and checking to use `time.Date`. The latter actually seemed to produce more code as I have to error check all the string conversions rather than just checking the error from `time.Parse`. 

I'd be happy to discuss this if I'm missing something but it didn't seem like I was gaining much simplicity when I tried to implement it.

Reply:
OK, this is what I came up with:
```
var blobNameRegex = regexp.MustCompile(`^(?:[^/]*/)?year=(\d{4})/month=(\d{2})/day=(\d{2})/hour=(\d{2})/(?:minute=(\d{2})/)?([^/].*)$`)

// parseBlobPath returns true if the blob is within the existing time range
func parseBlobPath(prefix *string, blobName string) (blobTime *time.Time, telemetryType component.DataType, err error) {
	matches := blobNameRegex.FindStringSubmatch(blobName)
	if matches == nil {
		err = errInvalidBlobPath
		return
	}

	year := matches[1]
	month := matches[2]
	day := matches[3]
	hour := matches[4]

	minute := "00"
	if matches[5] != "" {
		minute = matches[5]
	}

	lastPart := matches[6]

	timeString := fmt.Sprintf("%s-%s-%sT%s:%s", year, month, day, hour, minute)

	// Parse the expected format
	parsedTime, timeErr := time.Parse(timeFormat, timeString)
	if timeErr != nil {
		err = fmt.Errorf("parse blob time: %w", timeErr)
		return
	}
	blobTime = &parsedTime

	switch {
	case strings.Contains(lastPart, metricBlobSignifier):
		telemetryType = component.DataTypeMetrics
	case strings.Contains(lastPart, logsBlobSignifier):
		telemetryType = component.DataTypeLogs
	case strings.Contains(lastPart, tracesBlobSignifier):
		telemetryType = component.DataTypeTraces
	}

	return
}
```

This fails one of the tests, where a different error is returned (year=nope would not match the regex since nope is not digits), but otherwise it passes tests.

What I like about the regex solution is that it totally avoids that state-machine loop, which can be kinda hard to follow due to how the control flow jumps around, while this solution reads easily from top to bottom.

Reply:
Solution looks good. Yeah the test, needed to be tweaked to trigger the error it was targeting. I'm pushing up that change.